find_ideal_patchsize_and_magnification.py

```diff
@@ -0,0 +1,260 @@
+compare_plot(4, np.argmin(np.abs(np.array(areas)-np.percentile(areas, 1))), y_label = "Smallest | 1% | patch ")
+compare_plot(5, np.argmin(np.abs(np.array(areas)-np.percentile(areas, 5))), y_label = "Smallest | 5% | patch ")
+compare_plot(6, np.argmin(np.abs(np.array(areas)-np.percentile(areas, 10))), y_label = "Smallest | 10% | patch ")
+compare_plot(7, np.argmin(np.abs(np.array(areas)-np.percentile(areas, 15))), y_label = "Smallest | 15% | patch ")
+
+# 9th is average area, 10th is largest 5% by area 
+compare_plot(8, np.argmin(np.abs(np.array(areas)-np.percentile(areas, 50))), y_label = "Average | 50% | patch ")
+compare_plot(9, np.argmin(np.abs(np.array(areas)-np.percentile(areas, 95))), y_label = "Largest | 95% | patch ")
+
+plt.show()
+
+# +
+# Choose a final magnification and update patch metrics
+final_magnification = 20
+
+areas_final = areas*(final_magnification/initial_magnification)**2
+rect_final = rect_max*(final_magnification/initial_magnification)
+resize = final_magnification/initial_magnification
+centroids_final = centroids*(final_magnification/initial_magnification)
+
+# +
+# Find a patch size that works the best (Size Test)
+
+mark_patchsizes = [(512,"yellow"),(256,"green"),(128,"orange"),(64,"red")] # What patch sizes to test
+bins = 40 # Number of bins in histograms
+
+# Visualize patch areas compared to the largest circle that can fit in each patch size
+areas_total = areas_final[areas_final < 220000] # Filter outliers to make graphs readable
+a = plt.figure(figsize=(14,5))
+plt.subplot(1, 2, 1)
+plt.hist(areas_total,bins=bins)
+for ps,color in mark_patchsizes:
+    plt.axvline(x=math.pi*((ps/2)**2), color=color)
+plt.title("Areas")
+
+# Visualize patch maximum vertical/horizontal dimensions compared to each patch size
+rect_total = rect_final[rect_final < 600] # Filter outliers to make graphs readable
+plt.subplot(1, 2, 2)
+plt.hist(rect_total,bins=bins)
+for ps,color in mark_patchsizes:
+    plt.axvline(x=ps, color=color)
+plt.title("Normal Axis Lengths")
+plt.show()
+
+# Print the number of patches that will be cut off at each patch size
+#patchsizes = [x[0] for x in mark_patchsizes]
+mark_patchsizes.sort()
+for dim,_ in mark_patchsizes:
+    i = len(rect_final[rect_final<=dim])
+    print(f"{i} / {total_ooi}, {int(100*i/total_ooi)}% fit in {dim}px patch size")
+
+# +
+# Create a matrix of OOI that are cutoff at each patch size in mark_patchsizes, each column is a different patch size
+# A patch cutoff at a larger patch size in the mark_patchsizes will not be included in smaller patch sizes
+
+mask = True # Visualize with or without mask applied
+padding = 64 # Visualize at larger patch size to show what is cutoff
+example_count = 5 # Number of cutoff examples per patch size
+
+mark_patchsizes.sort(reverse=1)
+a = plt.figure(figsize=(15,20))
+plt.tight_layout(pad = 0.5)
+dim = (example_count,len(mark_patchsizes))
+rect_cutoff = rect_final.copy()
+
+
+for i,(ps,_) in enumerate(mark_patchsizes):
```

Comment:
similar here, what is the _ variable for?

Reply:
I wanted to keep the color and desired patch sizes together (to avoid mismatching and confusion). This for loop did not need the color.

Reply:
well organized!